pyat/at/lattice/lattice_object.py

```diff
@@ -314,21 +306,176 @@ def _addition_filter(self, elems: Iterable[Element]):
+                warn(AtWarning('Non-integer number of cells: {}/{}. Periodi'
+                               'city set to 1'.format(self.periodicity, n)))
+                periodicity = 1
+        try:
+            cell_h = self._cell_harmnumber
+        except AttributeError:
+            hdict = {}
+        else:
+            hdict = dict(_cell_harmnumber=n*cell_h)
+        elems = (copy_fun(el, copy_elements) for _ in range(n) for el in self)
+        return Lattice(elem_generator, elems, iterator=self.attrs_filter,
+                       periodicity=periodicity, **hdict)
+
+    def concatenate(self, *lattices: Iterable[Element],
+                    copy_elements=False, copy=False):
+        """Concatenate several `Iterable[Element]` with the lattice
+
+        Equivalents syntaxes:
+        >>> newring = ring.concatenate(r1, r2, r3, copy=True)
+        >>> newring = ring + r1 + r2 + r3
+
+        >>> ring.concatenate(r1, copy=False)
+        >>> ring += r1
+
+        Parameters:
+            lattices: :py:obj:`Iterables[Element]` to be concatenanted
+                      to the Lattice, several lattices are allowed
+                      (see example)
+
+        Keyword Arguments:
+            copy_elements(bool): Default :py:obj:`False`. If :py:obj:`True`
+                        deepcopies of the elements of lattices are used
+            copy(bool): Default :py:obj:`False`. If :py:obj:`True`
+                           the lattice is modified in place.
+                           Oterwise a new Lattice object is returned
+
+        Returns:
+            lattice(Lattice): concatenated Lattice, if `copy==True` the
+                              new lattice object is returned
+                              otherwise None
+        """
+        if copy:
+            lattice = Lattice(self)
+        else:
+            lattice = self
+        for lat in lattices:
+            lattice.extend(lat, copy_elements=copy_elements)
+        return lattice if copy else None
+
+    def reverse(self, copy=False):
+        r"""Reverse the order of the lattice and swapt the faces
+        of elements. Alignment errors are not swapped
+
+
+        Equivalents syntaxes:
+        >>> newring = ring.reverse(copy=True)
+        >>> newring = reversed(ring)
+
+        Keyword Arguments:
+            copy(bool): Default :py:obj:`False`. If :py:obj:`True`
+                           the lattice is modified in place.
+                           Oterwise a new Lattice object is returned
+
+        Returns:
+            lattice(Lattice): reversed Lattice, if `copy==True` the
+                              new lattice object is returned
+                              otherwise None
+        """
+        if copy:
+            elems = (el.swap_faces(copy=copy) for el in self[::-1])
+            return Lattice(elem_generator, elems, iterator=self.attrs_filter,
+                           periodicity=self.periodicity)            
```

Comment:
`__reversed__` still needs to be modified to return an iterator instead of a Lattice. It could be:
```python
def __reversed__(self):
    return (el.swap_faces(copy=True) for el in self[::-1])
```
Which could make this simpler:
```python
    if copy:
        return Lattice(elem_generator, reversed(self),
                       iterator=self.attrs_filter)
    else:
        reversed_list = list(reversed(self))
        self[:] = reversed_list
```
Specifying `periodicity` in the constructor is not necessary, the periodicity is transferred by `attrs_filter`